biosys/apps/main/api/serializers.py

```diff
@@ -0,0 +1,279 @@
+from __future__ import absolute_import, unicode_literals, print_function, division
+
+import datetime
+
+from django.core.exceptions import ValidationError
+from django.utils import timezone
+
+from rest_framework import serializers
+from rest_framework_bulk import BulkSerializerMixin, BulkListSerializer
+
+from main.models import Project, Site, Dataset, GenericRecord, Observation, SpeciesObservation
+from main.constants import MODEL_SRID
+
+
+class ProjectSerializer(serializers.ModelSerializer):
+    timezone = serializers.CharField()
+
+    class Meta:
+        model = Project
+        fields = '__all__'
+
+
+class SiteSerializer(BulkSerializerMixin, serializers.ModelSerializer):
+    class Meta:
+        model = Site
+        list_serializer_class = BulkListSerializer
+        fields = '__all__'
+
+
+class DatasetSerializer(serializers.ModelSerializer):
+    class DataPackageValidator:
+        def __init__(self):
+            self.dataset_type = Dataset.TYPE_GENERIC
+
+        def __call__(self, value):
+            Dataset.validate_data_package(value, self.dataset_type)
+
+        def set_context(self, serializer_field):
+            data = serializer_field.parent.context['request'].data
+            self.dataset_type = data.get('type')
+
+    data_package = serializers.JSONField(
+        validators=[
+            DataPackageValidator()
+        ]
+    )
+
+    class Meta:
+        model = Dataset
+        fields = '__all__'
+
+
+class GenericDataValidator:
+    def __call__(self, data):
+        if self.dataset is not None:
+            self.dataset.validate_data(data)
+            # validate site
+            schema = self.dataset.schema
+            site_fk = schema.get_fk_for_model('Site')
+            if site_fk:
+                model_field = site_fk.model_field
+                site_value = data.get(site_fk.data_field)
+                kwargs = {
+                    "project": self.dataset.project,
+                    model_field: site_value
+                }
+                site = Site.objects.filter(**kwargs).first()
+                if site is None:
+                    msg = "Could not find the site '{} in: {}':".format(site_value, data)
+                    raise ValidationError(msg)
+
+    def set_context(self, serializer_field):
```

Comment:
Yes it's a convention for  a Validator in DRF:
[http://www.django-rest-framework.org/api-guide/validators/#using-set_context](url)